src/main/scala/io/qbeast/spark/index/SparkPointWeightIndexer.scala

```diff
@@ -70,22 +34,26 @@ private class SparkPointWeightIndexer(tableChanges: TableChanges, isReplication:
   def buildIndex: DataFrame => DataFrame = (weightedDataFrame: DataFrame) => {
 
     val revision = tableChanges.updatedRevision
-    val replicatedSet = tableChanges.replicatedSet
-    val announcedSet = tableChanges.announcedSet
     val columnsToIndex = revision.columnTransformers.map(_.columnName)
-
-    weightedDataFrame
-      .withColumn(
-        cubeColumnName,
-        explode(
-          findTargetCubeIdsUDF(
-            struct(columnsToIndex.map(col): _*),
-            col(weightColumnName), {
-              if (isReplication) col(cubeToReplicateColumnName)
-              else lit(null)
-            })))
-      .transform(addState(columnsToIndex.length, announcedSet, replicatedSet))
-      .drop(cubeToReplicateColumnName)
+    if (isReplication) {
+      weightedDataFrame
+        .withColumn(
+          cubeColumnName,
+          explode(
+            findTargetCubeIdsUDF(
+              struct(columnsToIndex.map(col): _*),
+              col(weightColumnName),
+              col(cubeToReplicateColumnName))))
+    } else {
+      weightedDataFrame
+        .withColumn(
+          cubeColumnName,
+          explode(
+            findTargetCubeIdsUDF(
+              struct(columnsToIndex.map(col): _*),
+              col(weightColumnName),
+              lit(null))))
```

Comment:
I don't know which is the best workaround for separating replication and indexing at this stage. The options are:

1. Create different findTargetCubeIdsUDF with different arguments so we don't have to create a column when it's not necessary.
2. Create different SparkPointWeightIndexer implementations
3. Leave it as it is: a column with a null value is created
4. Other 

Any suggestion is welcome @eavilaes  @Jiaweihu08 @cugni !

Reply:
I think it is the case it is best to create 2 UDF. We haven't analyzed well enough in the optimize phase to understand the cost of adding columns and adding more logic, of if it is best to change the algorithm as a whole. Also, an udf without replication can simplify a bit the function logic, which is not bad.

Reply:
Yes, less overhead on creating a column and checking if it's null :+1: